clippy_utils/src/diagnostics.rs

```diff
@@ -36,6 +36,15 @@ fn docs_link(diag: &mut Diag<'_, ()>, lint: &'static Lint) {
 /// Usually it's nicer to provide more context for lint messages.
 /// Be sure the output is understandable when you use this method.
 ///
+/// NOTE: only lint-level attributes at the `LintPass::check_*` node from which you are calling this
+/// will be considered.
+/// This can be confusing if the given span is at a different place, because users won't know where
+/// `#[allow]` or `#[expect]` attributes need to be placed.
+///
+/// This can happen if, for example, you are in `LintPass::check_block` and you are emitting a lint
+/// for a particular expression within that block.
+/// In those cases, consider using [`span_lint_hir`], and pass the `HirId` of that expression.
+///
```

Comment:
What do you think about a formulation like this?

> NOTE: Rustc's lint emissions are always bound to a node in the HIR, which is used to determine the lint level. By default, the node that was passed into the `LintPass::check_*` function is used. If you're emitting the lint at the span of a different node than the one provided by the `LintPass::check_*` call, consider using [`span_lint_hir`] instead. This will enable `#[allow]` and `#[expect]` attributes to work on the expression highlighted in the displayed warning.
>
> If you're unsure, which function you should use, you can simply test if `#[allow]` attributes work everywhere, where you would expect them to work. Otherwise, [`span_lint_hir`] is probably the function you want to use.